### DISCUSSION TITLE
Custom module is not updating

### DISCUSSION THREAD

**Discussion by fredx100** [Alexays/Waybar#3046] [Q&A] [answered]
I've written a bash script to use within a custom module. I'm trying to use it as an infinite loop but the custom module never updates.

The script should display the alsa volume. It waits on `alsactl monitor` for an alsa event and then inspects the output of `amixer` to get the details. I'm very new to waybar so I've probably just got the config wrong or something elementary.

My script (`volume.sh`) seems fine from the terminal. (for reference:

```bash
#!/bin/bash

# Loop forever
while : ; do
    # Mangle the output of amixer into json
    amixer get Master |                                                                                                                                  
        grep 'Mono:' |
        sed --unbuffered -nre 's/.*\[off\].*/\{"text": "  M x", "class": "muted", "percentage": 0\}/p'\
                           -e 's/.*\[(.*)%\].*/\{"text": "\1%  ", "class": "unmuted", "percentage": \1\}/p'

    # Wait until the next alsa event
    /usr/sbin/alsactl monitor | grep -m 0 .

    # Exit if alsactl monitor was interrupted
    if (( ${PIPESTATUS[0]} != 141 )); then
            break;
    fi
done
```

) and my waybar config is

```yaml
"custom/alsa": {
    "exec": "~/.config/waybar/volume.sh",
    "return-type": "json",
    "on-click": "amixer set Master toggle",
    "on-scroll-up": "amixer set Master 5+",
    "on-scroll-down": "amixer set Master 5-",
    "tooltip": false
}
```

On launch of waybar the custom module displays correctly, but it never updates.

From a terminal <code>./volume.sh</code> outputs a line every time I change the volume or mute/unmute, which is what I want. The output from the terminal appears correct:

    $ ./volume.sh
    {"text": "43% ", "class": "unmuted", "percentage": 43}
    {"text": "50% ", "class": "unmuted", "percentage": 50}
    {"text": "  M x ", "class": "muted", "percentage": 0}
    {"text": "57% ", "class": "unmuted", "percentage": 57}

(this would be after 4 events) which, I believe is a sensible output for waybar. As such I expect the issue is with what form the output should take (this will be specific to waybar, potentially) or how the module is configured.

If I remove the loop from <code>volume.sh</code> and put an interval into my custom module config then the module updates correctly, but I'm hoping to avoid polling.

I'm aware that some scripts buffer when not outputting to a terminal and so I've tried prefixing the call to `volume.sh` with `stdbuf -oL`, which the internet says would force line buffering. I've also tried prefixing the pinelines within the bash script in the same way. Neither helped.

Any ideas about why my custom waybar module's not updating? Thanks!

**Comment by ordy** [comment]
Why not remove the loop, just have your amixer line in your script and add signal and interval to your config?

```bash
#!/bin/bash

amixer get Master | grep 'Mono:' | 
    sed --unbuffered -nre 's/.*\[off\].*/\{"text": "  M x", "class": "muted", "percentage": 0\}/p'\
                           -e 's/.*\[(.*)%\].*/\{"text": "\1%  ", "class": "unmuted", "percentage": \1\}/p'
```

```json
"custom/alsa": {
    "exec": "~/.config/waybar/volume.sh",
    "return-type": "json",
    "signal": 1,
    "interval": 1,
    "on-click": "amixer set Master toggle",
    "on-scroll-up": "amixer set Master 5+",
    "on-scroll-down": "amixer set Master 5-",
    "tooltip": false
}
```

**Comment by fredx100** [comment]
Yes, that would work but relies on polling. As stated in the original post, I'm hoping to avoid this. I appreciate that avoiding polling is a "want" rather than a "need" but I'm confused as to why the elegant solution (waiting on `alsactl monitor`) doesn't work. Thanks for the suggestion, though.

**Reply by ordy** [reply]
Oh my bad, completely missed the paragraph where you mentioned your were trying to avoid polling.

There is the option to not use interval and just signal, it would work on scroll events, but the value won't update if you change it from another place (like alsamixer, shortcuts or physical buttons)... So probably not ideal.

**Comment by fredx100** [comment, marked as answer]
I have a working solution and so can guess at reasons!

The working solution is

```
#!/bin/bash

set -e

createJson(){
    amixer get Master |
        grep 'Mono:' |
        sed -nre 's/.*\[(.*)%\].*\[off\].*/\{"text": "\1% x", "class": "muted", "percentage": \1\}/p'\
            -e   's/.*\[(.*)%\].*/\{"text": "\1%  ", "class": "unmuted", "percentage": \1\}/p'
}

createJson
while read line; do
    createJson
done < <( /usr/sbin/alsactl monitor )
```
This reads one line at a time of the output from `alsactl`, pumping the relevant info from `amixer` each time (and ignoring the line from `alsactl`). I also tweaked the sed line so that you could still see the volume when it was muted, but that's just cosmetic.

I expect this works because the pipe containing the output of `alsactl` is never broken. In the original script `grep` terminates after each line, breaking the pipeline. Somehow this detatched `alsactl` (sending a `SIGCHLD` which showed up in the debug output from waybar) and knackered the script (but only in non-interactive mode). In the newer script `alsactl` just keeps piping to the while loop until we quit.

**Comment by tienbuigia** [comment]
I have a similar problem with a script to display dwl's layout. Your mention of stdbuf helps me.

```
  "custom/dwl-layout": {
    "format": " {}",
    // "exec": "dwlmsg -wl | while read -r layout; do echo $layout | sed 's/.* //'; done"
    // "exec": "dwlmsg -wl"
    // "exec": "dwlmsg -wl | sed 's/.* //'"
    "exec": "dwlmsg -wl | stdbuf -oL sed 's/.* //'"
  },
```

This is my simple script using dwlmsg, it watches dwl layout (ipc). The while loop works, without sed works, but using sed will not output anything on waybar. It does work on terminal:

```
❯ dwlmsg -wl | sed 's/.* //'
[]=
[3]
[]=
^C
```

And then I add `stdbuf -oL` as you mention and it works nicely now. You may need to try again, this time add `stdbuf` to the `grep`.
